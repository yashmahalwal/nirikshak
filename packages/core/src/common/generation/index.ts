export { generateFaker } from "./fakerGen";
export { generateCustom } from "./customGen";
