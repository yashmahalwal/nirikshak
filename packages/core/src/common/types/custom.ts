// Structuring similar to faker types
// Custom function string is path to a helper function kept in the working directory

import { Literal } from "./literals";

// Custom function string, starts with custom:
// Ex: custom:random.name
export type CustomFunctionString = string;

// Object format of the custom function
// Ex: {function: "custom:random.name"}
export type CustomFunctionObject = {
    function: CustomFunctionString;
    args?: any[];
};

// A custom function that user provides
// User needs to guarantee the validity at runtime
export interface CustomFunction {
    // Primitive is simply a literal or an array of nested arrays of literals
    // Ex: 1, [1,7], [false, [true, [7,8,false, [null]]]]
    (...args: any[]): Promise<Literal>;
}
// Custom type
export type CustomFunctionType = CustomFunctionString | CustomFunctionObject;

// type guard: CustomFunctionString
export function isCustomFunctionString(
    input: any
): input is CustomFunctionString {
    if (typeof input !== "string") return false;

    return input.startsWith("custom:");
}

// type guard: CustomFunctinObject
export function isCustomFunctionObject(
    input: any
): input is CustomFunctionObject {
    if (!input || typeof input !== "object") return false;

    return (
        "function" in input &&
        isCustomFunctionString(input["function"]) &&
        ("args" in input ? Array.isArray(input["args"]) : true)
    );
}

// type guard: CustomFunction
export function isCustomFunction(input: any): input is CustomFunctionType {
    return isCustomFunctionString(input) || isCustomFunctionObject(input);
}

// Normalized form of custom functions
// Same as CustomFunctionObject but optional fields made compulsort
// Ex: {function : "custom:random.name", args: []}
export type NormalizedCustomFunction = {
    [K in keyof CustomFunctionObject]-?: CustomFunctionObject[K];
};

// Function to normalize custom functions
export function normalizeCustomFunction(
    input: CustomFunctionType
): NormalizedCustomFunction {
    if (isCustomFunctionObject(input)) return { args: [], ...input };
    return {
        function: input,
        args: [],
    };
}
