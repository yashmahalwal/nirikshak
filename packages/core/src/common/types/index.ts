export { SchemaHelpers } from "./helpers";
export { isCustomFunction, CustomFunction } from "./custom";
export { isFakerType } from "./fakerTypes";
