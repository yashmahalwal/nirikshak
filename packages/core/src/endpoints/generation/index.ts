export { generateURL } from "./urlStringGen";
export { getFromResource } from "./resourceStringGen";
export * from "./helpers";
export * from "./input";
