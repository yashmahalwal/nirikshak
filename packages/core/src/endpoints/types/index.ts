export {
    MethodType,
    HeaderMap,
    Cases,
    BodyType,
    BodyInstance,
} from "./helpers";
export { isURLString } from "./urlString";
export { isResourceString } from "./resourceString";
export { isDescription, Description } from "./description";
export { HeaderAndQuery, Inputs } from "./input";
export { HeaderAndStatus, Outputs } from "./output";
