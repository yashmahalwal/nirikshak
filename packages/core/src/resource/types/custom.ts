import { Literal } from "./primitive";

// User need to guarantee the validity at runtime
export type Custom = string;
export interface CustomFunction {
  (): Literal;
}
