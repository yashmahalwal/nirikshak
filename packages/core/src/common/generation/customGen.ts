import { CustomFunctionType, normalizeCustomFunction } from "../types/custom";
import _ from "lodash";
import { Literal, isLiteral } from "../types/literals";
import { SchemaHelpers } from "../types/helpers";

// Function to extract the custom function from helper and generate value
export async function generateCustom(
    input: CustomFunctionType,
    helpers: SchemaHelpers
): Promise<Literal> {
    const object = normalizeCustomFunction(input);
    let value: Promise<any>;
    try {
        value = _.get(helpers, object.function.slice(7))(...object.args);
    } catch (e) {
        throw new Error(
            `function: ${object.function}, args: ${object.args.join(",")}`
        );
    }

    const resolvedValue = await value;
    if (!isLiteral(resolvedValue))
        throw new Error(`Non literal type value from ${object.function}`);
    return resolvedValue;
}
