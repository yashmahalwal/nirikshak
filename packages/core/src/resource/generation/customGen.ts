import {
    CustomFunctionType,
    normalizeCustomFunction,
    CustomFunction,
} from "../types/custom";
import { ResourceHelpers } from "../types/helper";
import _ from "lodash";

// Function to extract the custom function from helper and generate value
export function generateCustom(
    input: CustomFunctionType,
    helpers: ResourceHelpers
): ReturnType<CustomFunction> {
    const object = normalizeCustomFunction(input);
    try {
        return _.get(helpers, object.function.slice(7))(...object.args);
    } catch (e) {
        throw new Error(
            `function: ${object.function}, args: ${object.args.join(",")}`
        );
    }
}
