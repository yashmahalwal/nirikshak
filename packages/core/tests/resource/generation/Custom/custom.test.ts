import { CustomFunctionType } from "../../../../src/resource/types/custom";
import { ResourceHelpers } from "../../../../src/resource/types/helper";
import faker from "faker";
import { generateCustom } from "../../../../src/resource/generation/customGen";

// Helpers for the resource
const Helpers: ResourceHelpers = {
  username: async () => faker.name.firstName() + faker.name.lastName(),
  number: async (min: number, max: number, step: number) =>
    // Use faker.random.number instead of Math.random
    // For universal randomness seed
    faker.random.number({ min, max, precision: step }),
};

// Valid custom function types
const ValidCustoms: CustomFunctionType[] = [
  "custom:username",
  {
    function: "custom:number",
    args: [{ min: 0, max: 1, precision: 0.1 }],
  },
];

describe("Custom literal generation", () => {
  // Settings randomness seed
  // Does not work with faker.random.uuid and faker.date.*
  beforeAll(() => faker.seed(123));

  test.each(ValidCustoms)(`Valid custom: %#`, async (entry) =>
    expect(await generateCustom(entry, Helpers)).toMatchSnapshot()
  );

  // Checking that the function throws an error if no valid helper found at runtime
  test("Invalid custom literal: No helper found", () =>
    expect(
      generateCustom("custom:lipsum", Helpers)
    ).rejects.toMatchInlineSnapshot(
      `[Error: function: custom:lipsum, args: ]`
    ));
});
