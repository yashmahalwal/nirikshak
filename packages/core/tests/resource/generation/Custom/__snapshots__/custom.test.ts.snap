// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom literal generation Valid custom: 0 1`] = `"MaurineRath"`;

exports[`Custom literal generation Valid custom: 1 1`] = `NaN`;
